### ISSUE TITLE
Performance issue when extracting large archives.

### ISSUE THREAD

**Issue by isoos** [simolus3/tar#14]
Example archive: https://storage.googleapis.com/pub-packages/packages/fhir-0.4.0.tar.gz
Iterating through the entries and getting their content is really slow on a relatively modern CPU. Possibly buffering / buffer sizing / buffer copies are at fault here.

**Comment by simolus3** [comment]
Thanks for the report! Can you post how you go iterate though the files? Are you just calling `hasNext()` and then write the contents to a file?

**Comment by jonasfj** [comment]
Hmm, I can't reproduce...
```dart
import 'dart:io';
import 'package:tar/tar.dart';

Future<void> main() async {
  final u =
      'https://storage.googleapis.com/pub-packages/packages/fhir-0.4.0.tar.gz';
  final c = HttpClient();
  final req = await c.getUrl(Uri.parse(u));
  final res = await req.close();
  final reader = TarReader(res.transform(gzip.decoder));

  while (await reader.moveNext()) {
    final header = reader.current.header;
    var size = 0;
    await reader.current.contents.forEach((chunk) {
      size += chunk.length;
    });
    print('${header.name}: $size bytes');
  }
  await reader.cancel();
  c.close();
}
```

**Comment by isoos** [comment]
```
void main() async {
  final rs = await http.get(Uri.parse(
      'https://storage.googleapis.com/pub-packages/packages/fhir-0.4.0.tar.gz'));
  final decompressed = gzip.decode(rs.bodyBytes);
  final reader = TarReader(Stream.fromIterable([decompressed]));
  final sw = Stopwatch()..start();
  while (await reader.moveNext()) {
    final entry = reader.current;
    final content = await entry.contents.toList();
    print(
        '${sw.elapsed} ${entry.header.name} ${entry.header.size} ${content.length}');
  }
}
```

```
0:00:00.789303 CHANGELOG.md 6456 1
0:00:01.930241 LICENSE 1080 1
0:00:03.067631 README.md 13657 1
0:00:04.183780 build.yaml 592 1
0:00:05.329544 example/lib/main.dart 3145 1
0:00:06.472146 example/pubspec.yaml 262 1
0:00:07.633339 lib/dstu2.dart 74 1
0:00:08.806767 lib/dstu2/basic_types/basic_types.dart 53 1
0:00:09.978639 lib/dstu2/basic_types/element.dart 1234 1
0:00:11.162848 lib/dstu2/basic_types/element.freezed.dart 6918 1
```

**Comment by jonasfj** [comment]
hmm, if everything is a single chunk then I bet `ChunkedStreamReader` might do something slow..

**Comment by jonasfj** [comment]
```dart
        if (_buffer.isNotEmpty) {
          if (size < _buffer.length) {
            final output = _buffer.sublist(0, size); // this is probably slow
            _buffer = _buffer.sublist(size);         // probably also slow...
            size = 0;
            yield output;
            _reading = false;
            break;
          }
```

**Comment by jonasfj** [comment]
Good news in `pub` we do streaming consumption... so this shouldn't be a problem :D

**Comment by isoos** [comment]
Yeah, sublist is super-slow. A relevant part from `package:buffer` for building a view:
https://github.com/isoos/buffer/blob/master/lib/buffer.dart#L423-L426

**Comment by simolus3** [comment]
@isoos As a temporary workaround, you can make the operation much faster by splitting the streams into chunks of 512:

```dart
extension on List<int> {
  Stream<Uint8List> get chunksForTar {
    const blockSize = 512;
    final buffer = this is Uint8List
        ? (this as Uint8List).buffer
        : Uint8List.fromList(this).buffer;
    final amountOfBlocks = (length + 1) ~/ blockSize;

    return Stream.fromIterable(Iterable.generate(amountOfBlocks, (i) {
      final start = i * blockSize;
      final blockLength = min(blockSize, length - start);
      return buffer.asUint8List(start, blockLength);
    }));
  }
}
```

**Comment by isoos** [comment]
Yeah, that could work, but in the end I've updated the download and extraction code here: https://github.com/dart-lang/pana/pull/883

**Comment by isoos** [comment]
Also checked the other use-cases in pub-dev, and we are doing the parsing with a stream from `File.openRead()` which should be fine.

**Comment by simolus3** [comment]
I tried to benchmark this specific scenario some more, by downloading `fhir-0.4.0.tar.gz` and deflating it to a tar file. I then ran 50 iterations with the current implementation, the faster chunked stream implementation and a custom chunking implementation specialized on fixed-length blocks.
Here is the median of 50 iterations (which is way faster than running it once, probably due to JIT kicking in):

| Implementation  | Stream from tar file | Read single chunk |
| ------------- | ------------- | ------------- |
| Current tar / current async  | `1087ms`  | _too long to test_ |
| Current tar / PR from Jonas  | `725ms`  | `327ms` |
| [Custom in tar](https://github.com/simolus3/tar/blob/3095ce75bf6beb649ebaa9e5f8ef25857d8139eb/lib/src/utils.dart#L240-L555)  | `508ms`  | `134ms` |
| `package:archive`  | _unsupported_  | `98ms` |

Almost all of the difference between the fastest `package:tar` and `package:archive` is spent verifying checksums (not done by `archive`) and detecting advanced headers, so I'm ok with that.
I think the takeaway here is that async IO is slow ([dartbug.com/44006](https://dartbug.com/44006)), I reckon that we'll be plenty fast if the chunked stream reader also supports specializations for `Uint8List`.

**Comment by simolus3** [comment]
This has been fixed in `tar` version `0.5.1`.
